erdjs/src/dapp/walletProvider.ts

```diff
@@ -0,0 +1,237 @@
+                return resolve(data.data);
+            };
+
+            window.addEventListener('message', isConnected);
+        });
+    }
+
+    /**
+     * Fetches the login hook url and redirects the client to the wallet login.
+     */
+    async login(): Promise<string> {
+        if (!this.mainFrame) {
+            return '';
+        }
+
+        const {contentWindow} = this.mainFrame;
+        if (!contentWindow) {
+            return '';
+        }
+
+        return new Promise<string>((resolve, reject) => {
+            contentWindow.postMessage({
+                type: DAPP_MESSAGE_CONNECT_URL,
+            }, this.walletUrl);
+
+            const timeout = setTimeout(_ => reject('connect url not responding'), 5000);
+            const connectUrl = (ev: IDappMessageEvent) => {
+                if (!this.isValidWalletSource(ev.origin)) {
+                    return;
+                }
+
+                const {data} = ev;
+                if (data.type !== DAPP_MESSAGE_CONNECT_URL) {
+                    return;
+                }
+
+                clearTimeout(timeout);
+                window.removeEventListener('message', connectUrl.bind(this));
+                return resolve(data.data.toString());
+            };
+
+            window.addEventListener('message', connectUrl);
+        }).then((connectionUrl: string) => {
+            const redirectParts = connectionUrl.split('=');
+            const redirectAddress = redirectParts[redirectParts.length - 1];
+            if (redirectParts.length < 2 || !redirectAddress.startsWith('erd1')) {
+
+            }
+            window.location.href = `${this.baseWalletUrl()}${connectionUrl}?callbackUrl=${window.location.href}/dashboard`;
+            return window.location.href;
+        }).catch(_ => {
+            return '';
+        });
+    }
+
+    /**
+     * Packs a {@link Transaction} and fetches correct redirect URL from the wallet API. Then redirects
+     *   the client to the send transaction hook
+     * @param transaction
+     */
+    async sendTransaction(transaction: Transaction): Promise<Transaction> {
+        if (!this.mainFrame) {
+            throw new Error("Wallet provider is not initialised, call init() first");
+        }
+
+        const {contentWindow} = this.mainFrame;
+        if (!contentWindow) {
+            throw new Error("Wallet provider is not initialised, call init() first");
+        }
+
+        return new Promise<Transaction>((resolve, reject) => {
+            contentWindow.postMessage({
+                type: DAPP_MESSAGE_SEND_TRANSACTION_URL,
+                data: {
+                    transaction: {
+                        transaction
+                    }
+                }
+            }, this.walletUrl);
+
+            const timeout = setTimeout(_ => reject('send transaction url not responding'), 5000);
+            const sendTransactionUrl = (ev: IDappMessageEvent) => {
```

Comment:
Can we extract `sendTransactionUrl` to a separate, private function?

Reply:
Same as above ^